microsoft/testsuites/core/boot.py

```diff
@@ -0,0 +1,86 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+from pathlib import Path
+
+from assertpy.assertpy import assert_that
+
+from lisa import (
+    Logger,
+    RemoteNode,
+    TestCaseMetadata,
+    TestSuite,
+    TestSuiteMetadata,
+    simple_requirement,
+)
+from lisa.features import SerialConsole
+from lisa.operating_system import Redhat
+from lisa.tools import Uname
+from lisa.util import LisaException, SkippedException
+from lisa.util.shell import wait_tcp_port_ready
+
+
+@TestSuiteMetadata(
+    area="core",
+    category="functional",
+    description="""
+    This test suite is to test VM working well after updating on VM and rebooting.
+    """,
+)
+class Boot(TestSuite):
+    @TestCaseMetadata(
+        description="""
+        This test case will
+        1. Skip testing if the distro is not redhat type.
+        2. Install kernel-debug package and set boot with this debug kernel.
+        3. Reboot VM, check kernel version is debug type.
+        """,
+        priority=3,
+        requirement=simple_requirement(
+            supported_features=[SerialConsole],
+        ),
+    )
+    def boot_with_debug_kernel(
+        self, log: Logger, node: RemoteNode, log_path: Path
+    ) -> None:
+        # 1. Skip testing if the distro is not redhat type.
+        if not isinstance(node.os, Redhat):
+            raise SkippedException(
+                f"{node.os.name} not supported. "
+                "This test case only supports Redhat distros."
+            )
+
+        # 2. Install kernel-debug package and set boot with this debug kernel.
+        node.os.install_packages("kernel-debug")
+        result = node.execute("grub2-set-default 0", sudo=True)
+        result.assert_exit_code()
+
+        # 3. Reboot VM, check kernel version is debug type.
+        try:
+            node.reboot()
+        except Exception as identifier:
+            serial_console = node.features[SerialConsole]
+            # if there is any panic, fail before partial pass
+            serial_console.check_panic(saved_path=log_path, stage="reboot")
+
+            # if node cannot be connected after reboot, it should be failed.
+            if isinstance(identifier, LisaException) and str(identifier).startswith(
+                "cannot connect to TCP port"
+            ):
+                raise LisaException(f"after reboot, {identifier}")
+        is_ready, tcp_error_code = wait_tcp_port_ready(
```

Comment:
Because this logic appears in every reboot place. We need to move them into the reboot method.

Reply:
if use remotenode in reboot, will encounter circular import issue.
```
ImportError: cannot import name 'RemoteNode' from partially initialized module 'lisa' (most likely due to a circular import) (C:\CODE\internal\main\lsg-lisa\lisa\lisa\__init__.py)
```

Reply:
Refer to `uname.py` or other classes, which import Node. Import it in TYPE_CHECKING, and use it with quotes in declaration. Unless the type is used in `isinstance`, otherwise, it should be ok.